src/main/scala/is/hail/variant/Call.scala

```diff
@@ -3,113 +3,318 @@ package is.hail.variant
 import java.io.Serializable
 
 import is.hail.check.Gen
+import is.hail.expr.Parser
 import is.hail.utils._
 
+import scala.annotation.switch
+import scala.collection.JavaConverters._
+import scala.language.implicitConversions
+
+object Call0 extends Serializable {
+  def apply(phased: Boolean = false): Call = {
+    Call(0, phased, ploidy = 0)
+  }
+}
+
+object Call1 extends Serializable {
+  def apply(aj: Int, phased: Boolean = false): Call = {
+    require(aj >= 0, s"allele index must be >= 0. Found $aj.")
+    Call(aj, phased, ploidy = 1)
+  }
+}
+
+object Call2 extends Serializable {
+  def fromUnphasedDiploidGtIndex(gt: Int): Call = {
+    require(gt >= 0, s"gt must be >= 0. Found $gt.")
+    Call(gt, phased = false, ploidy = 2)
+  }
+
+  def apply(aj: Int, ak: Int, phased: Boolean = false): Call = {
+    require(aj >= 0 && ak >= 0, s"allele indices must be >= 0. Found j=$aj and k=$ak.")
+
+    val alleleRepr =
+      if (phased)
+        Genotype.diploidGtIndex(aj, aj + ak)
+      else
+        Genotype.diploidGtIndexWithSwap(aj, ak)
+
+    Call(alleleRepr, phased, ploidy = 2)
+  }
+}
+
+object CallN extends Serializable {
+  def apply(alleles: java.util.ArrayList[Int], phased: Boolean): Call = apply(alleles.asScala.toArray, phased)
+
+  def apply(alleles: Array[Int], phased: Boolean = false): Call = {
+    val ploidy = alleles.length
+    (ploidy: @switch) match {
+      case 0 => Call0(phased)
+      case 1 => Call1(alleles(0), phased)
+      case 2 => Call2(alleles(0), alleles(1), phased)
+      case _ => throw new UnsupportedOperationException
+    }
+  }
+}
+
 object Call extends Serializable {
+  def apply(ar: Int, phased: Boolean, ploidy: Int): Call = {
+    require(ploidy >= 0 && ploidy <= 2, s"invalid ploidy: $ploidy. Only support ploidy in range [0, 2]")
```

Comment:
we can't do this -- it allocates a bunch of strings an an anonymous thunk class to fulfil the `=> T` argument of `require`. if you want to throw nice error messages here, you'll have to do it with 
```
if (!(...))
    throw new InvalidArgumentException(...)
```
which will allocate only if the exception is raised.

Reply:
It's a total deficiency of Scala.